packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -721,6 +738,89 @@ function createFunctionComponentUpdateQueue(): FunctionComponentUpdateQueue {
   };
 }
 
+function use<T>(usable: Usable<T>): T {
+  if (
+    usable !== null &&
+    typeof usable === 'object' &&
+    typeof usable.then === 'function'
+  ) {
+    // This is a thenable.
+    const thenable: Thenable<T> = (usable: any);
+
+    // Track the position of the thenable within this fiber.
+    const index = thenableIndexCounter;
+    thenableIndexCounter += 1;
+
+    switch (thenable.status) {
+      case 'fulfilled': {
+        const fulfilledValue: T = thenable.value;
+        return fulfilledValue;
+      }
+      case 'rejected': {
+        const rejectedError = thenable.reason;
+        throw rejectedError;
+      }
+      default: {
+        const prevThenableAtIndex: Thenable<T> | null = getSuspendedThenable(
+          index,
+        );
+        if (prevThenableAtIndex !== null) {
+          switch (prevThenableAtIndex.status) {
+            case 'fulfilled': {
+              const fulfilledValue: T = prevThenableAtIndex.value;
+              return fulfilledValue;
+            }
+            case 'rejected': {
+              const rejectedError: mixed = prevThenableAtIndex.reason;
+              throw rejectedError;
+            }
+            default: {
+              // The thenable still hasn't resolved. Suspend with the same
+              // thenable as last time to avoid redundant listeners.
+              throw prevThenableAtIndex;
```

Comment:
This seems like it should be an error case - e.g. you used something mutable conditionally.

We shouldn't bother retrying immediately unless `lastUsedThenable` was pinged. There's no point in retrying due to anything else because that can no longer be blocking. Anything preceding lastUsedThenable must have already been resolved or we wouldn't have gotten that far.

If something else pings maybe we need to reset the suspended thenables anyway - e.g. if we want to treat it as a restart.

So if we have anything in the slot from before, it must've been resolved. Otherwise something went wrong.

Reply:
Yeah that seems reasonable. I started a list of a warning cases but maybe this one should be an actual error

Reply:
> We shouldn't bother retrying immediately unless lastUsedThenable was pinged. There's no point in retrying due to anything else because that can no longer be blocking. Anything preceding lastUsedThenable must have already been resolved or we wouldn't have gotten that far.

Yeah that is how it currently works. It's a different variable though because it also works for ad-hoc wakeables (ones that aren't `use`-d, but the same logic you're describing:

https://github.com/facebook/react/blob/796d31809b3683083d3b62ccbab4f00dec8ffb1f/packages/react-reconciler/src/ReactFiberWakeable.new.js#L28-L34